engine/access/state_stream/subscription.go

```diff
@@ -0,0 +1,115 @@
+package state_stream
+
+import (
+	"context"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// GetDataByHeightFunc is a callback used by subscriptions to retrieve data for a given height.
+// Expected errors:
+// - storage.ErrNotFound
+// - execution_data.BlobNotFoundError
+// All other errors are considered exceptions
+type GetDataByHeightFunc func(ctx context.Context, height uint64) (interface{}, error)
+
+// Subscription represents a streaming request, and handles the communication between the grpc handler
+// and the backend implementation.
+type Subscription interface {
+	// ID returns the unique identifier for this subscription used for logging
+	ID() string
+
+	// Channel returns the channel from which subscriptino data can be read
+	Channel() <-chan interface{}
+
+	// Err returns the error that caused the subscription to fail
+	Err() error
+}
+
+type SubscriptionImpl struct {
+	id string
+
+	// ch is the channel used to pass data to the receiver
+	ch chan interface{}
+
+	// err is the error that caused the subscription to fail
+	err error
+}
+
+func NewSubscription() *SubscriptionImpl {
+	return &SubscriptionImpl{
+		id: uuid.New().String(),
+		ch: make(chan interface{}),
```

Comment:
nit: non-buffered channels are prone to deadlocks.

Reply:
I can add a buffer, but this is intentionally setup without one to apply backpressure. Send uses a timeout to fail and close the connection if the client is slow
https://github.com/onflow/flow-go/blob/6397de16ba4323d474ea5a3c6d5d9996ed7a2694/engine/access/state_stream/subscription.go#L78-L88

Reply:
i think we can balance perf and backpressure by having a tiny buffer-- that way we'll pipeline fetching with sending.  (It should be already somewhat pipelined by async nature of Send, but we may still be blocked by tcp/h2 flow control)

Reply:
that's fair. what would you say is an appropriate buffer size? 5-10?

Reply:
we can start with 1 and then go up